### PR TITLE
Features: Core IAM module, X-ray submodule

### LINKED ISSUE #2: Feature: Roles / Groups for admin, developer, CI
This is all in the existing source CloudFormation stuff, but basically we should re-examine what we're doing with the following in mind:

- [ ] Identify specific capabilities for admin, developer, ci
- [ ] Consider what's the difference between developer and ci as they're currently the same.
- [ ] Fully document roles / groups for admin, developer, ci
- [ ] Document potential "gotcha's" such as having an admin create _new_ resources in a related serverless (really CloudFormation) config and then needing an admin to delete them, where ci role wouldn't be able to do this. (Can separately ticket this one if needed as it's complicated -- Roemer has some specific experiences in mind).
- [ ] Implement IAM permissions for each.

### LINKED ISSUE #4: [DOCUMENT] Feature: Stack / resource naming.
There's often confusion in where resources come from when using serverless (which creates CloudFormation stacks) vs other things support that like this process.

In the reference app, we have specific naming prefixes like:

- `"sls-${ServiceName}-${Stage}"`: In `serverless.yml` as `service:` value.
- `"aws-${ServiceName}-${Stage}"`: In AWS CloudFormation.

For our new project we should:

- [ ] Support the variables `ServiceName` and `Stage`
- [ ] Create a base Terraform name of `"tf-${ServiceName}-${Stage}"` to signal that these are Terraform managed.
    - [ ] Possibly also include the name of this project? (We could even give it a "cool", short name instead of the super boring `serverless-iam-terraform`)
    - [ ] Document this scheme.
- [ ] In documentation, suggest users name their serverless app `service`  `"sls-${ServiceName}-${Stage}"` with potentially an example.

### LINKED ISSUE #7: CI: Set up
- [ ] `terraform fmt` + failure if changes (?)

### LINKED ISSUE #9: Feature/Option: Xray
Dependencies:

- IAM

```yml
# TODO(IamPolicyAdmin): Xray: view traces
- Effect: Allow
  Action:
  - xray:BatchGetTraces
  - xray:GetServiceGraph
  - xray:GetTraceGraph
  - xray:GetTraceSummaries
  Resource:
  # Must be wildcard.
  # https://docs.aws.amazon.com/IAM/latest/UserGuide/list_awsxray.html
  # https://docs.aws.amazon.com/xray/latest/devguide/xray-permissions.html#xray-permissions-managedpolicies
  - "*"
# TODO(IamPolicyLambdaExecution): Xray: upload traces
- Effect: Allow
  Action:
  - xray:PutTraceSegments
  - xray:PutTelemetryRecords
  Resource:
  # Must be wildcard.
  # https://docs.aws.amazon.com/IAM/latest/UserGuide/list_awsxray.html
  # https://docs.aws.amazon.com/xray/latest/devguide/xray-permissions.html#xray-permissions-managedpolicies
  - "*"
```

### LINKED ISSUE #14: [Release] Rename project
Should be available on both Terraform and AWS.

Potential `serverless-aws` (?)

### LINKED ISSUE #16: [Release][Docs] Root README.md
- [x] Installation / Usage
- [ ] Add note about minimum TF version.
- [ ] Secrets: AWS_PROFILE or AWS_*_KEYS docs
- [x] Pairing and using with `serverless` framework
- [x] Link / discussion to reference app.
- [x] The IAM roles approach. Discuss admin, developer, ci
- [x] Dedicated, hierarchical section on "scenarios"
- [ ] Extending and enhancing: (I’m thinking about customizations for this. If I just export the IAM developer|ci|admin groups and the IAM lambda execution role as outputs, is that sufficient for your use upstream assuming you had a dependency on this new “sls IAM permissions” TF module? Then, you’d likely enhance with `aws_iam_group_policy_attachment` for the groups and `aws_iam_role_policy_attachment` for the lambda execution role…)

### LINKED ISSUE #17: [Release][Docs] Per-module/submodule READMEs
- [ ] IAM/Root
- [ ] Search and update this ticket with all others.
- [ ] Make sure to check how this looks in TF registry.

### LINKED ISSUE #20: [Release][Docs] CONTRIBUTING.md document
- [ ] How to publish the package.
- [ ] ... everything else

### PR DESCRIPTION
**Pull request by ryan-roemer**
### Overview

This (massive) PR is intended to be the initial release of:

- The core IAM module to support a `serverless` framework application
- An optional X-Ray submodule to add IAM permissions needed for `serverless` X-ray support.

The companion reference application PR to review is: https://github.com/FormidableLabs/aws-lambda-serverless-reference/pull/7

Things I'd suggest to review (if all the files are too beastly):

- Core README: https://github.com/FormidableLabs/terraform-aws-serverless/blob/feature/xray/README.md
- X-ray README: https://github.com/FormidableLabs/terraform-aws-serverless/blob/feature/xray/modules/xray/README.md
- The variables (and all the ARNs to limit gathered in one place): https://github.com/FormidableLabs/terraform-aws-serverless/blob/feature/xray/variables.tf

### Sample Integration

Basic, simplest integration is like this:

```hcl
# Base `serverless` IAM support.
module "serverless" {
  source = "FormidableLabs/serverless/aws"

  region       = "${var.region}"
  service_name = "${var.service_name}"
  stage        = "${var.stage}"
}

# OPTION(Xray): Add X-ray support to lambda execution roles.
# **NOTE**: No explicit dependency on core, but it **is** needed.
module "serverless_xray" {
  source = "FormidableLabs/serverless/aws//modules/xray"

  region       = "${var.region}"
  service_name = "${var.service_name}"
  stage        = "${var.stage}"
}
```

### Work

Completed:

- Add CONTRIBUTING.md document. Fixes #20
- Per-module/submodule READMEs. Fixes #17
- Root README.md. Fixes #16
- Rename project. Fixes #14
- CI: Set up. Fixes #7
- Feature: Stack / resource naming. Fixes #4
- Feature: Roles / Groups for admin, developer, CI enhancement. Fixes #2
- Feature/Option: Xray enhancement. Fixes #9

In progress:

- Review/remove all source TODOs. #15
- Publish: Terraform Registry. #5 

/cc @tptee @jasonwilson @kevinmstephens @mscottx88 @jjasonclark @ianwsperber

### CODE REVIEW

File: HISTORY.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+History
```
Comment (tptee) [inline HISTORY.md]: Nit: is `HISTORY.md` more prevalent than `CHANGELOG.md`? Maybe worth changing depending on what our big projects use...

File: README.md [hunk 2]
```diff
@@ -1,24 +1,158 @@
...
+- [Integration](#integration)
+  - [Reference project](#reference-project)
+  - [Module integration](#module-integration)
+  - [AWS IAM group integration](#aws-iam-group-integration)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Getting a [serverless][] application all the way to production in AWS **correctly** using **securely** can be quite challenging. In particular, things like:
```
Comment (tptee) [inline README.md]: Nit: should `using` be `and` in this sentence?

File: README.md [hunk 3]
```diff
@@ -1,24 +1,158 @@
...
+- **IAM Groups**: Role-specific groups to attach to AWS users to give humans and CI the minimum level of permissions locked to both a specific `serverless` service and stage/environment.
+
+## Concepts
+
+This module allows practical isolation / compartmentalization of privileges within a single AWS account along the folowing axes:
+
+* **Stage/Environment**: An arbitrary environment to isolate -- this module doesn't restrict selection in any way other than there has to be at least one. In practice, a good set of choices may be something like `sandbox`, `development`, `staging`, `production`.
+* **IAM Groups**: This module creates/enforces a scheme wherein:
+    * **Admin**: AWS users assigned to the `admin` group can create/update/delete a `serverless` application and do pretty much anything that the `serverless` framework permits out of the box.
+    * **Developer, CI**: AWS users assigned to the `developer|ci` groups can update a `serverless` application and do non-mutating things like view logs, perform rollbacks, etc.
```
Comment (tptee) [inline README.md]: Nit: `read-only` might be more clear than `non-mutating`

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  group      = "${aws_iam_group.ci.name}"
+  policy_arn = "${aws_iam_policy.developer.arn}"
+}
+
+# developer
+resource "aws_iam_group" "developer" {
+  name = "${local.tf_service_name}-${local.stage}-developer"
+}
+
+resource "aws_iam_group_policy_attachment" "developer_developer" {
```
Comment (tptee) [inline main.tf]: image

File: modules/xray/variables.tf [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  # **NOTE**: This is difficult to lock down because we need the actual
+  # `$api-id` for a `!Ref` which we don't know ahead of time. It's
+  # created dynamically as a part of `sls` provisioning.
+  #
+  # TODO(8): Research locking down more.
+  # https://github.com/FormidableLabs/serverless-iam-terraform/issues/8
+  #
+  # See:
+  # - https://iam.cloudonaut.io/reference/apigateway.html
+  # - https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html
```
Comment (tptee) [inline modules/xray/variables.tf]: For future reference, you might be able to lock this down with resource policies (I've done so for a private endpoint but I think it works for any endpoint type): https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-resource-policies-create-attach.html

File: HISTORY.md (hunk 1 above)
Comment (ryan-roemer) [reply HISTORY.md]: Yeah, I think `CHANGELOG.md` is more common for the TF modules I'm seeing. I'll change it.

File: README.md (hunk 2 above)
Comment (ryan-roemer) [reply README.md]: Yeah, I'm `using` the wrong word 😛

File: README.md (hunk 3 above)
Comment (ryan-roemer) [reply README.md]: For this one, `rollback` is more than read-only. And it actually somewhat mutates too...

Maybe just replace with `do things like`...?

File: modules/xray/variables.tf (hunk 5 above)
Comment (ryan-roemer) [reply modules/xray/variables.tf]: Great note! Added to #8

File: README.md (hunk 3 above)
Comment (tptee) [reply README.md]: Oh! Yeah that works!

File: policy-ci.tf [hunk 6]
```diff
@@ -0,0 +1,7 @@
+###############################################################################
+# Policy: CI
+# ----------
+# A CI user has all the privileges of a developer and presently has no
+# additional permissions, so this file remains empty for future use.
+###############################################################################
+
```
Comment (ianwsperber) [inline policy-ci.tf]: I wonder if it  might be better to define the policy with empty permissions, just to encourage people to  add CI  specific permissions should they encounter  them

Comment (ryan-roemer) [reply policy-ci.tf]: That's what I originally wanted to do and had! Unfortunately, having an empty policy blows up during `apply` (although weirdly not during `plan`).

Comment (ryan-roemer) [reply policy-ci.tf]: This is what I originally tried (it's the removal commit): https://github.com/FormidableLabs/terraform-aws-serverless/pull/22/commits/c1ac9713eb3c60fa3155a99827dd7ae19bdcf27c

If you've got other ideas, I'm game!

**Comment by ryan-roemer** [comment]
I'm merging so I can figure out how to publish, etc., but post-merge review most welcome!

**Pull request merged**